hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferContainer.java

```diff
@@ -0,0 +1,342 @@
+                break;
+            }
+        }
+
+        return seq;
+    }
+
+    /**
+     * Cleans up the ringbuffer by deleting all expired items.
+     */
+    public void cleanup() {
+        if (!isRetentionEnabled() || headSequence > tailSequence) {
+            return;
+        }
+
+        long now = currentTimeMillis();
+        while (headSequence <= tailSequence) {
+            int index = toIndex(headSequence);
+
+            if (ringExpirationMs[index] > now) {
+                return;
+            }
+
+            // we null the slot and allow the gc to take care of the object.
+            // if we don't clean it, we'll have a potential memory leak.
+            ringItems[index] = null;
+
+            // we don't need to 0 the ringExpirationMs slot since it contains a long value.
+
+            // and we move the head to the next item.
+            // if nothing remains in the ringbuffer, then the head will be 1 larger than the tail.
+            headSequence++;
+        }
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeLong(tailSequence);
+        out.writeLong(headSequence);
+        out.writeInt(capacity);
+        out.writeLong(retentionMillis);
+        out.writeInt(inMemoryFormat.ordinal());
+
+        boolean retentionEnabled = isRetentionEnabled();
+
+        long now = System.currentTimeMillis();
+
+        // we only write the actual content of the ringbuffer. So we don't write empty slots.
+        for (long seq = headSequence; seq <= tailSequence; seq++) {
+            int index = toIndex(seq);
+
+            if (inMemoryFormat == BINARY) {
+                out.writeData((Data) ringItems[index]);
+            } else {
+                out.writeObject(ringItems[index]);
+            }
+
+            // we write the time difference compared to now. Because the clock on the receiving side
+            // can be totally different then our time. If there would be a retention of 10 seconds, than
+            // the expiration time would be 10 seconds after the insertion time. But the if ringbuffer is
+            // migrated to a machine with a time 1 hour earlier, the retention would effectively become
+            // 1 hours and 10 seconds.
+            if (retentionEnabled) {
+                long deltaMs = ringExpirationMs[index] - now;
+                out.writeLong(deltaMs);
+            }
+        }
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        tailSequence = in.readLong();
+        headSequence = in.readLong();
+        capacity = in.readInt();
+        retentionMillis = in.readLong();
+        inMemoryFormat = values()[in.readInt()];
+
+        ringItems = new Object[capacity];
+
+        boolean retentionEnabled = isRetentionEnabled();
+        if (retentionEnabled) {
+            ringExpirationMs = new long[capacity];
+        }
+
+        long now = System.currentTimeMillis();
```

Comment:
If you use `com.hazelcast.util.Clock`, it will be easier to test the time difference you described above.